File: Data Structures & Algorithms/longest-increasing-subsequence/submission-6.java
class Solution {
    private int[][] dp;
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        dp = new int[n][n+1];
        for(int[] row : dp){
            Arrays.fill(row, -1);
        }
        return helper(nums, 0, -1);
    }
    public int helper(int[] nums, int i, int prev){
        if (i == nums.length){
            return 0;
        }
        if (dp[i][prev+1] != -1){
            return dp[i][prev+1];
        }
        int res = helper(nums, i+1, prev); // No take
        if (prev == -1 || nums[prev] < nums[i]){
            res = Math.max(res, 1 + helper(nums, i+1, i)); // Take
        }
        dp[i][prev+1] = res;
        return res;
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-6.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [-1] * (n+1)
        def helper(i):
            if i < 0:
                return 0
            if i == 0:
                return 1
            if dp[i] != -1:
                return dp[i]
            
            dp[i] = helper(i-1) + helper(i-2)
            return dp[i]
        helper(n)
        return dp[n]